### PR TITLE
Fix custom filters to correctly handle hashtag filters

### LINKED ISSUE #27990: Custom Filters incorrectly match for whole word keyword filters when written as a "#hashtag"
### Steps to reproduce the problem

1. Create a custom filter for `#example` and apply the Whole Word option
2. Create a status with the text `"Test filter https://example.org/#example test"`
4. Fetch the custom filters with `CustomFilter.cached_filters_for("<account id>")`
5. Apply the filters to the created status: `CustomFilter.apply_cached_filters(filters, status)`

### Expected behaviour

The filter should not match because "https://example.org/#example" is not including "#example" at a word boundary

### Actual behaviour

Filter matches the hash fragment in the URL

### Detailed description

It seems this bug is introduced by the logic of the whole word filter application:

```
sb = /\A[[:word:]]/.match?(keyword.keyword) ? '\b' : ''
eb = /[[:word:]]\z/.match?(keyword.keyword) ? '\b' : ''

/(?mix:#{sb}#{Regexp.escape(keyword.keyword)}#{eb})/
```

Where `/\A[[:word:]]/` does not match on symbols (e.g., it'll return false for "#example")

This results in the word boundaries not being respected.

### Mastodon instance

_No response_

### Mastodon version

main

### Technical details

If this is happening on your own Mastodon server, please fill out those:
- Ruby version: 3.2.2
- Node.js version: 20.9

### PR DESCRIPTION
**Pull request by ThisIsMissEm**
The previous logic didn't match keywords that started with hashtags correctly. Fixes #27990.

### CODE REVIEW

File: app/models/custom_filter.rb [hunk 1]
```diff
@@ -70,8 +70,8 @@ def self.cached_filters_for(account_id)
       scope.to_a.group_by(&:custom_filter).each do |filter, keywords|
         keywords.map! do |keyword|
           if keyword.whole_word
-            sb = /\A[[:word:]]/.match?(keyword.keyword) ? '\b' : ''
-            eb = /[[:word:]]\z/.match?(keyword.keyword) ? '\b' : ''
+            sb = /\A|^[[:word:]]/.match?(keyword.keyword) ? '\b' : ''
+            eb = /[[:word:]]\z|$/.match?(keyword.keyword) ? '\b' : ''
```
Comment (ThisIsMissEm) [inline app/models/custom_filter.rb:74]: Whilst this does work, we may actually want to use the logic that the streaming server uses here, which is "If the filter starts with or end with non-whitespace, do a boundary match"

```suggestion
            sb = /^[\S]/.match?(keyword.keyword) ? '\b' : ''
            eb = /[\S]$/.match?(keyword.keyword) ? '\b' : ''
```

It seems ruby's \A and \z do not match symbols, only alphanumerical characters.

File: spec/requests/api/v1/statuses_spec.rb [hunk 2]
```diff
@@ -26,16 +26,20 @@
       end
 
       context 'when post includes filtered terms' do
```
Comment (ThisIsMissEm) [inline spec/requests/api/v1/statuses_spec.rb:28]: This is likely not the most ideal way to test this, but it was one of the tests I could find for custom filters, there didn't actually seem to be tests directly for custom filters.

**Comment by ThisIsMissEm** [comment]
> With this change, `https://foo/bar#filter_me` would still match, because there is a word boundary between `bar` and `#filter`.

That's not what the tests seem to show? Granted, I just cobbled this PR together from trying stuff out in irb and the tests.. so maybe there's a better solution?

**Comment by ClearlyClaire** [comment]
> That's not what the tests seem to show? Granted, I just cobbled this PR together from trying stuff out in irb and the tests.. so maybe there's a better solution?

The tests you added feature `https://example.com/#filter_me`, which is a different situation, as there is no word boundary between `https://example.com/` and `#filter_me`, as neither `/` nor `#` are word characters.

**Comment by ThisIsMissEm** [comment]
> > That's not what the tests seem to show? Granted, I just cobbled this PR together from trying stuff out in irb and the tests.. so maybe there's a better solution?
> 
> The tests you added feature `https://example.com/#filter_me`, which is a different situation, as there is no word boundary between `https://example.com/` and `#filter_me`, as neither `/` nor `#` are word characters.

Ah, so you're saying the code would still match hash fragments in URLs where the path component doesn't include a trailing slash?

I'll try to add more tests in a bit. Any thoughts on testing the custom filter model directly?

**Comment by ClearlyClaire** [comment]
> Ah, so you're saying the code would still match hash fragments in URLs where the path component doesn't include a trailing slash?

Yes. I'm not completely sure what's the expectation is and how to actually achieve it though.

**Comment by ThisIsMissEm** [comment]
@ClearlyClaire there was another suggestion I saw about the ability to mute hashtags, maybe that'd work as an additional option instead of reusing the text filtering?

Then we could prohibit filters starting with #

**Comment by ClearlyClaire** [comment]
> @ClearlyClaire there was [another suggestion](https://github.com/mastodon/mastodon/discussions/21762) I saw about the ability to mute hashtags, maybe that'd work as an additional option instead of reusing the text filtering?

Why not, this could also make sense with out-of-band hashtags.

> Then we could prohibit filters starting with #

I don't see the point in prohibiting such filters.

**Comment by ThisIsMissEm** [comment]
> > Then we could prohibit filters starting with #
> 
> I don't see the point in prohibiting such filters.

Well, they don't work as one likely expects them to, so having the keyword filters instead be just word characters might make sense? i.e., we'd be nudging that you actually want a hashtag filter.

> > @ClearlyClaire there was [another suggestion](https://github.com/mastodon/mastodon/discussions/21762) I saw about the ability to mute hashtags, maybe that'd work as an additional option instead of reusing the text filtering?
> 
> Why not, this could also make sense with out-of-band hashtags.

Given this, I think closing this PR is the correct answer, and we'll make sure to have a ticket to implement custom hashtag filters.

**Pull request closed without merging**